### PR TITLE
✨(frontend) Prevent to purchase a product

### PR DESCRIPTION
**Pull request by rlecellier**
## Purpose

Product should not be available for sell when it has no remaining order left.

### payment button

image

### payment modal

image

### CODE REVIEW

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx [hunk 1]
```diff
@@ -70,15 +81,20 @@ const PurchaseButton = ({ product, disabled }: PurchaseButtonProps) => {
...
           >
             {product.call_to_action}
-          </button>
+          </Button>
           {!hasAtLeastOneCourseRun && (
             <p className="product-item__no-course-run">
               <FormattedMessage {...messages.noCourseRunToPurchase} />
             </p>
           )}
+          {!hasAtLeastOneRemainingOrder && (
```
Comment (rlecellier) [inline src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx:93]: We may replace this by : 
```js
{hasAtLeastOneCourseRun && !hasAtLeastOneRemainingOrder && (
```
it depend if we want both messages and if not, what message do we want to prioritize.

Comment (jbpenrath) [reply src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx:93]: I think we want first to display feedback message to inform there is one target course without course run (`hasAtLeastOneCourseRun`) and second display feedback message to inform there isn't available seat (`!hasAtLeastOneRemainingOrder`)

File: src/frontend/js/utils/test/factories/joanie.ts [hunk 2]
```diff
@@ -108,6 +108,7 @@ export const CertificateProductFactory = factory((): Product => {
     certificate_definition: CertificationDefinitionFactory().one(),
     orders: [],
     target_courses: TargetCourseFactory().many(5),
+    remaining_order_count: 12,
```
Comment (jbpenrath) [inline src/frontend/js/utils/test/factories/joanie.ts:111]: We could use a random value between 1 and 100 ?

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx [hunk 3]
```diff
@@ -17,6 +18,12 @@ const messages = defineMessages({
     description: "Label displayed inside the product's when there is no courseRun",
     id: 'components.SaleTunnel.noCourseRunToPurchase',
   },
+  noRemainingOrder: {
+    defaultMessage: 'This product is full, is currently not available for sale',
+    description:
+      "Label displayed inside the product's when there is no remaining order for the product",
+    id: 'components.SaleTunnel.noRemainingOrder',
+  },
```
Comment (jbpenrath) [inline src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx:26]:
```suggestion
  noRemainingOrder: {
    defaultMessage: 'There are no more places available for this course run.',
    description:
      "Label displayed below the product's CTA when there is no remaining available seat for the product",
    id: 'components.SaleTunnel.noRemainingOrder',
  },
```

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx [hunk 4]
```diff
@@ -56,11 +63,15 @@ const PurchaseButton = ({ product, disabled }: PurchaseButtonProps) => {
...
+  const hasAtLeastOneRemainingOrder =
+    product.remaining_order_count && product.remaining_order_count > 0;
+  const isPurchasable = hasAtLeastOneRemainingOrder && hasAtLeastOneCourseRun;
+
   return (
     <>
       {!disabled && (
         <>
-          <button
+          <Button
```
Comment (jbpenrath) [inline src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx:74]: Isn't there visual issue to use Cunningham button on the course syllabus as other course run enrollment button are old ones ?

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.spec.tsx [hunk 5]
```diff
@@ -74,14 +74,37 @@ describe('PurchaseButton', () => {
...
-    await act(async () => {
+    act(() => {
       fireEvent.click(button);
     });
 
     // - SaleTunnel should have been opened
     screen.getByTestId('SaleTunnel__modal');
   });
 
+  it('renders a disabled CTA if the product have no remaining orders', async () => {
```
Comment (jbpenrath) [inline src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.spec.tsx:85]: Can you also write a test to ensure that all is working as expected when `remaining_order_count` is `null`.

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx (hunk 4 above)
Comment (rlecellier) [reply src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx:74]: I thought that we want to migrate all buttons to cunningham buttons, so i did this one on the way.
Should i reverse my change ?

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx (hunk 1 above)
Comment (jbpenrath) [reply src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx:93]: I would display only one message at a time with the priority I described above.

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx (hunk 4 above)
Comment (jbpenrath) [reply src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.tsx:74]: Yup, we will make this change in a dedicated PR as we have several things to check.

File: src/frontend/js/api/joanie.ts [hunk 6]
```diff
@@ -35,20 +46,14 @@ export function checkStatus(
...
-    return response.text();
+    return getResponseBody(response);
   }
 
   if (options.ignoredErrorStatus.includes(response.status)) {
     return Promise.resolve(options.fallbackValue);
   }
 
-  throw new HttpError(response.status, response.statusText);
+  throw new HttpError(response.status, response.statusText, undefined, getResponseBody(response));
```
Comment (jbpenrath) [inline src/frontend/js/api/joanie.ts:56]: Good move ! Can you just create a test suite for this method to ensure all is working fine when response body has no content please.

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.spec.tsx [hunk 7]
```diff
@@ -73,13 +76,68 @@ describe('PurchaseButton', () => {
...
     // Then user can enter into the sale tunnel and follow its 3 steps
-    await act(async () => {
-      fireEvent.click(button);
+    await act(async () => userEvent.click(button));
+
+    // - SaleTunnel should have been opened
+    expect(screen.getByTestId('SaleTunnel__modal')).toBeInTheDocument();
+  });
+
+  it.only('shows cta to open sale tunnel when remaning orders is null', async () => {
```
Comment (jbpenrath) [inline src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.spec.tsx:87]:
```suggestion
  it('shows cta to open sale tunnel when remaning orders is null', async () => {
```

File: src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.spec.tsx [hunk 8]
```diff
@@ -73,13 +76,68 @@ describe('PurchaseButton', () => {
...
+    const product = ProductFactory({ remaining_order_count: null }).one();
```
Comment (jbpenrath) [inline src/frontend/js/widgets/CourseProductItem/components/PurchaseButton/index.spec.tsx:88]: Can you also test that the `hasAtLeastOneCourseRun` has priority over `hasAtLeastOneRemainingOrder`.

**Review by jbpenrath — APPROVED** [review]
Just two comments then merge it
